plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java

```diff
@@ -51,30 +62,92 @@
-    private final ConcurrentMap<DatasetId, DatasetId> datasetIds = new ConcurrentHashMap<>();
+    private final boolean caseInsensitiveNameMatching;
+    private final Cache<String, String> remoteDatasetNames;
+    private final Cache<String, Map<String, String>> remoteTableNames;
 
     BigQueryClient(BigQuery bigQuery, BigQueryConfig config)
     {
         this.bigQuery = bigQuery;
         this.viewMaterializationProject = config.getViewMaterializationProject();
         this.viewMaterializationDataset = config.getViewMaterializationDataset();
+        Duration caseInsensitiveNameMatchingCacheTtl = requireNonNull(config.getCaseInsensitiveNameMatchingCacheTtl(), "caseInsensitiveNameMatchingCacheTtl is null");
+
+        this.caseInsensitiveNameMatching = config.isCaseInsensitiveNameMatching();
+        CacheBuilder<Object, Object> remoteNamesCacheBuilder = CacheBuilder.newBuilder()
+                .expireAfterWrite(caseInsensitiveNameMatchingCacheTtl.toMillis(), MILLISECONDS);
+        this.remoteDatasetNames = remoteNamesCacheBuilder.build();
+        this.remoteTableNames = remoteNamesCacheBuilder.build();
     }
 
-    TableInfo getTable(TableId tableId)
+    String toRemoteDatasetName(String projectId, String datasetName)
     {
-        TableId bigQueryTableId = tableIds.get(tableId);
-        Table table = bigQuery.getTable(bigQueryTableId != null ? bigQueryTableId : tableId);
-        if (table != null) {
-            tableIds.putIfAbsent(tableId, table.getTableId());
-            datasetIds.putIfAbsent(toDatasetId(tableId), toDatasetId(table.getTableId()));
+        requireNonNull(projectId, "projectId is null");
+        requireNonNull(datasetName, "datasetName is null");
+        verify(CharMatcher.forPredicate(Character::isUpperCase).matchesNoneOf(datasetName), "Expected schema name from internal metadata to be lowercase: %s", datasetName);
+        if (!caseInsensitiveNameMatching) {
+            return datasetName;
+        }
+
+        try {
+            @Nullable String remoteDatasetName = remoteDatasetNames.getIfPresent(datasetName);
+            if (remoteDatasetName == null) {
+                // this might have been just created, reload cache
+                Map<String, String> mapping = StreamSupport.stream(listDatasets(projectId).spliterator(), false)
+                        // will throw on collisions to avoid ambiguity
+                        .collect(toImmutableMap(dataset -> dataset.getDatasetId().getDataset().toLowerCase(ENGLISH), dataset -> dataset.getDatasetId().getDataset()));
+                remoteDatasetNames.putAll(mapping);
+            }
+            remoteDatasetName = remoteDatasetNames.getIfPresent(datasetName);
+            if (remoteDatasetName != null) {
+                return remoteDatasetName;
+            }
+            // TODO: what to throw if still not found?
+            throw new TrinoException(BIGQUERY_OBJECT_NOT_FOUND, "Failed to find remote dataset name: " + datasetName);
```

Comment:
return original, like https://github.com/trinodb/trino/blob/91e0f05fcd9adafac80f35d840fb9796fd3c72cc/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java#L868

Reply:
But why? Is the expectation that it will fail somewhere deeper in the stack?

Reply:
Yes. And also so that it 'degrades gracefully' in case we have some issue in the code.
Anyway, let's keep it consistent with BaseJdbcClient, unless we have a good reason not to